core/src/test/java/io/grpc/inprocess/InProcessTransportTestWithCause.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2020 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.inprocess;
+
+import static org.junit.Assert.assertEquals;
+
+import io.grpc.Status;
+import io.grpc.internal.InternalServer;
+import io.grpc.internal.ManagedClientTransport;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class InProcessTransportTestWithCause extends InProcessTransportTest {
```

Comment:
Hmm... Extending creates a lot of unnecessary tests.

Let's create a new test from scratch. I'd basically start with [HelloWorldServerTest](https://github.com/grpc/grpc-java/blob/master/examples/src/test/java/io/grpc/examples/helloworld/HelloWorldServerTest.java), with some replacements. We don't want to deal with generated code here, so we can use [voidMethod()](https://grpc.github.io/grpc-java/javadoc/io/grpc/testing/TestMethodDescriptors.html#voidMethod--) and ClientCalls and implement the stub stuff ourselves.

```java
@Rule
public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();

...
final Throwable t = new Throwable();
ServerServiceDefinition def = ServerServiceDefinition.builder("Failing")
    .addMethod(TestMethodDescriptors.voidMethod(), new ServerCallHandler<Void, Void>() {
      @Override public ServerCall.Listener<Void> startCall(ServerCall<Void, Void> call, Metadata headers) {
        call.close(Status.UNAVAILABLE.withDescription("best error message").withCause(t), new Metadata());
        return new ServerCall.Listener() {};
      }
    })
    .build();
grpcCleanup.register(InProcessServerBuilder.forName("failing-server").directExecutor().addService(def).build().start());
Channel channel = grpcCleanup.register(InProcessChannelBuilder.forName("failing-server").directExecutor().build());
try {
  ClientCalls.blockingUnaryCall(channel, TestMethodDescriptors.voidMethod(), CallOptions.DEFAULT);
  fail("expected exception");
} catch (StatusRuntimeException ex) {
  assertEquals(t, ex.getCause());
}
```

I've just "dumped" that out there; feel free to change formatting and organization. I also don't claim it is 100% correct :smile:. But it should be close.

I know there are a lot of pieces there, but they are all virtually boilerplate. The only "interesting" things are creating the Throwable, call.close(), and catching StatusRuntimeException (which is how the blocking stub reports errors).

Reply:
I followed your advice with the test case. I made some small changes to it, but by and large it worked great! 

Organization-wise, I moved the new test case into the `InProcessTransportTest` class because I thought that it belonged there. But for this to work, I had to expose the field `server` in `AbstractTransportTest`: 
https://github.com/grpc/grpc-java/blob/2efad75c7b0895419d8465a1a410acb8c0b28569/core/src/test/java/io/grpc/internal/AbstractTransportTest.java#L158-L161
so that I could set `server=null`. I had to do this because the after each hook in `AbstractTransportTest` would try to call `server.shutdown` if I didnt (which would fail since I dont start that server in the test case).
https://github.com/grpc/grpc-java/blob/2efad75c7b0895419d8465a1a410acb8c0b28569/core/src/test/java/io/grpc/internal/AbstractTransportTest.java#L230-L242

So I'm not sure whether this was the right move. That being said I'm happy to change it! 

And thank you so much for all the help and pointers you've been giving me 🙂